File: MController/src/main/java/tk/munditv/mcontroller/dmr/MOSMediaPlayers.java

package tk.munditv.mcontroller.dmr;

import android.content.Context;
import android.util.Log;

import org.fourthline.cling.model.types.UnsignedIntegerFourBytes;
import org.fourthline.cling.support.lastchange.LastChange;
import org.fourthline.cling.support.model.TransportState;

import java.util.concurrent.ConcurrentHashMap;

/**
 * @author offbye
 */
public class MOSMediaPlayers extends ConcurrentHashMap<UnsignedIntegerFourBytes, MOSMediaPlayer> {

    private final static String TAG = MOSMediaPlayers.class.getSimpleName();

    final protected LastChange avTransportLastChange;
    final protected LastChange renderingControlLastChange;
    final protected LastChange tvControlLastChange;

    private Context mContext;


    public MOSMediaPlayers(int numberOfPlayers,
                           Context context,
                           LastChange avTransportLastChange,
                           LastChange renderingControlLastChange,
                           LastChange tvControlLastChange) {
        super(numberOfPlayers);
        Log.d(TAG, "MOSMediaPlayers()");

        this.mContext = context;
        this.avTransportLastChange = avTransportLastChange;
        this.renderingControlLastChange = renderingControlLastChange;
        this.tvControlLastChange = tvControlLastChange;

        for (int i = 0; i < numberOfPlayers; i++) {

            MOSMediaPlayer player =
                    new MOSMediaPlayer(
                            new UnsignedIntegerFourBytes(i),
                            mContext,
                            avTransportLastChange,
                            renderingControlLastChange,
                            tvControlLastChange
                    ) {
                        @Override
                        protected void transportStateChanged(TransportState newState) {
                            super.transportStateChanged(newState);
                            if (newState.equals(TransportState.PLAYING)) {
                                onPlay(this);
                            } else if (newState.equals(TransportState.STOPPED)) {
                                onStop(this);
                            }
                        }
                    };
            put(player.getInstanceId(), player);
        }
    }

    protected void onPlay(MOSMediaPlayer player) {
        Log.d(TAG, "Player is playing: " + player.getInstanceId());
    }

    protected void onStop(MOSMediaPlayer player) {
        Log.d(TAG, "Player is stopping: " + player.getInstanceId());
    }
}

File: TVService-kt/src/main/java/tk/munditv/mtvservice/dmr/MOSMediaPlayers.kt
package tk.munditv.mtvservice.dmr

import android.content.Context
import android.util.Log
import org.fourthline.cling.model.types.UnsignedIntegerFourBytes
import org.fourthline.cling.support.lastchange.LastChange
import org.fourthline.cling.support.model.*
import java.util.concurrent.ConcurrentHashMap

/**
 * @author offbye
 */
open class MOSMediaPlayers(numberOfPlayers: Int,
                           context: Context,
                           avTransportLastChange: LastChange,
                           renderingControlLastChange: LastChange,
                           tvControlLastChange: LastChange) : ConcurrentHashMap<UnsignedIntegerFourBytes?, MOSMediaPlayer?>(numberOfPlayers) {
    protected val avTransportLastChange: LastChange
    protected val renderingControlLastChange: LastChange
    protected val tvControlLastChange: LastChange
    private val mContext: Context
    protected open fun onPlay(player: MOSMediaPlayer) {
        Log.d(TAG, "Player is playing: " + player.instanceId)
    }

    protected open fun onStop(player: MOSMediaPlayer) {
        Log.d(TAG, "Player is stopping: " + player.instanceId)
    }

    companion object {
        private val TAG = MOSMediaPlayers::class.java.simpleName
    }

    init {
        Log.d(TAG, "MOSMediaPlayers()")
        mContext = context
        this.avTransportLastChange = avTransportLastChange
        this.renderingControlLastChange = renderingControlLastChange
        this.tvControlLastChange = tvControlLastChange
        for (i in 0 until numberOfPlayers) {
            val player: MOSMediaPlayer = object : MOSMediaPlayer(
                    UnsignedIntegerFourBytes(i*1L),
                    mContext,
                    avTransportLastChange,
                    renderingControlLastChange,
                    tvControlLastChange
            ) {
                override fun transportStateChanged(newState: TransportState) {
                    super.transportStateChanged(newState)
                    if (newState == TransportState.PLAYING) {
                        onPlay(this)
                    } else if (newState == TransportState.STOPPED) {
                        onStop(this)
                    }
                }
            }
            put(player.instanceId, player)
        }
    }

}